### PR TITLE
feat(aapd-861): link sqluser to appeal

### PR DESCRIPTION
**Pull request by bryn-pins**
## Ticket Number

https://pins-ds.atlassian.net/browse/AAPD-861

## Description of change

Link user and appeal when feature flag is on and user confirms email

## Checklist

- [ ] Requires infrastructure changes
- [ ] I have updated the documentation accordingly
- [x] My commit history in this PR is linear
- [x] New features have tests
- [ ] Breaking change (team conversation required)

## Important

Please do not merge from `main` (please only [rebase](https://github.com/foundry4/appeal-planning-decision/wiki/An-intro-to-Git-Rebase)). This keeps the history linear and easier to debug.

### CODE REVIEW

File: packages/forms-web-app/src/lib/appeals-api-wrapper.js [hunk 1]
```diff
@@ -276,6 +277,38 @@ exports.getUserAppealsById = async (id) => {
...
+ * @typedef {import('appeals-service-api').Api.AppealCaseWithAppellant} AppealCaseWithAppellant
+ */
+
+/**
+ * @param {string} email
+ * @param {string} appealSqlId
+ * @param {string} [role]
+ * @returns {Promise<AppealCaseWithAppellant>}
+ */
+exports.linkUserToV2Appeal = async (email, appealSqlId, role) => {
```
Comment (beejjacobs) [inline packages/forms-web-app/src/lib/appeals-api-wrapper.js:290]: Thoughts on a class-based approach for v2 API calls?

e.g. https://github.com/Planning-Inspectorate/appeal-planning-decision/pull/3334/files#diff-9727715fc7c722a6ac0258c1eddf1a7e9445e6bad6047711d142f81066bc8b6c

**Review by beejjacobs — COMMENTED** [review]
Looks good and thorough to me!

My only real question is: will this update the role if it changes during the journey? I thought we might need something on other parts of the journey or to check appeal submission update data?

**Comment by bryn-pins** [comment]
> Looks good and thorough to me!
> 
> My only real question is: will this update the role if it changes during the journey? I thought we might need something on other parts of the journey or to check appeal submission update data?

No, will do that in part two

File: packages/forms-web-app/src/lib/appeals-api-wrapper.js (hunk 1 above)
Comment (bryn-pins) [reply packages/forms-web-app/src/lib/appeals-api-wrapper.js:290]: Definitely preferable, want me to wait for that?

Comment (bryn-pins) [reply packages/forms-web-app/src/lib/appeals-api-wrapper.js:290]: I'll bring a copy in

**Review by beejjacobs — APPROVED** [review]

File: packages/common/src/client/appeals-api-client.js [hunk 2]
```diff
@@ -0,0 +1,172 @@
+const { default: fetch, AbortError } = require('node-fetch');
+const uuid = require('uuid');
```
Comment (beejjacobs) [inline packages/common/src/client/appeals-api-client.js:2]: don't need a package for this, we can use `crypto.randomUUID()`

File: packages/common/src/client/appeals-api-client.js [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+		const endpoint = `${v2}/users/${email}`;
+		const response = await this.#makeGetRequest(endpoint);
+		return response.json();
+	}
+
+	/**
+	 * Handles error responses and timeouts from calls to appeals api
+	 * @param {string} path endpoint to call e.g. /api/v2/users
+	 * @param {'GET'|'POST'|'PUT'|'DELETE'} [method] - request method, defaults to 'GET'
+	 * @param {object|undefined} [opts] - options to pass to fetch can include request body
```
Comment (beejjacobs) [inline packages/common/src/client/appeals-api-client.js:62]: FYI `[]` (optional) implies `|undefined`

File: packages/common/src/config.js
Comment (beejjacobs) [inline packages/common/src/config.js:1]: Appreciate it's already a pattern but I don't like using `process.env` in packages/libraries. Better to explicitly pass in any parameters that are required - which I think the ApiClient supports for `baseUrl`, we should pass in `timeout` too.

**Review by beejjacobs — COMMENTED** [review]
api client looks great! Couple of comments.

**Review by beejjacobs — APPROVED** [review]

**Pull request merged**